File: sdks/ios/Sources/RajutechieStreamKit/Modules/MeetingModule.swift
import Foundation

public struct Meeting: Codable {
    public let id: String
    public let title: String
    public let meetingCode: String
    public let status: String
    public let hostId: String
}

public struct MeetingParticipant: Codable {
    public let id: String
    public let meetingId: String
    public let userId: String
    public let role: String
    public let status: String
}

public struct MeetingConfigBuilder {
    private var title: String
    private var description: String?
    private var scheduledAt: Date?
    private var waitingRoom: Bool = true
    private var maxParticipants: Int = 100

    public init(title: String, scheduledAt: Date? = nil) {
        self.title = title
        self.scheduledAt = scheduledAt
    }

    public func waitingRoom(_ enabled: Bool) -> MeetingConfigBuilder {
        var copy = self
        copy.waitingRoom = enabled
        return copy
    }

    public func maxParticipants(_ count: Int) -> MeetingConfigBuilder {
        var copy = self
        copy.maxParticipants = count
        return copy
    }
}

public final class MeetingModule {
    private let http: HTTPClient
    private let ws: WebSocketClient

    init(http: HTTPClient, ws: WebSocketClient) {
        self.http = http
        self.ws = ws
    }

    public func schedule(_ config: MeetingConfigInput) async throws -> Meeting {
        let result: Meeting = try await http.post("/meetings", body: config)
        return result
    }

    public func get(_ meetingId: String) async throws -> Meeting {
        let result: Meeting = try await http.get("/meetings/\(meetingId)")
        return result
    }

    public func join(_ meetingId: String) async throws -> MeetingParticipant {
        let result: MeetingParticipant = try await http.post("/meetings/\(meetingId)/join")
        return result
    }

    public func joinByCode(_ code: String) async throws -> Meeting {
        let result: Meeting = try await http.get("/meetings/join/\(code)")
        return result
    }

    public func leave(_ meetingId: String) async throws {
        let _: EmptyResponse = try await http.post("/meetings/\(meetingId)/leave")
    }

    public func end(_ meetingId: String) async throws {
        let _: EmptyResponse = try await http.post("/meetings/\(meetingId)/end")
    }

    public func muteParticipant(_ meetingId: String, userId: String) async throws {
        let _: EmptyResponse = try await http.post("/meetings/\(meetingId)/participants/\(userId)/mute")
    }

    public func removeParticipant(_ meetingId: String, userId: String) async throws {
        let _: EmptyResponse = try await http.post("/meetings/\(meetingId)/participants/\(userId)/remove")
    }

    public func muteAll(_ meetingId: String) async throws {
        let _: EmptyResponse = try await http.post("/meetings/\(meetingId)/mute-all")
    }

    public func createPoll(_ meetingId: String, question: String, options: [String], isAnonymous: Bool = false) async throws -> MeetingPoll {
        struct CreatePollBody: Encodable {
            let question: String
            let options: [String]
            let isAnonymous: Bool
        }
        let result: MeetingPoll = try await http.post("/meetings/\(meetingId)/polls", body: CreatePollBody(question: question, options: options, isAnonymous: isAnonymous))
        return result
    }

    public func votePoll(_ meetingId: String, pollId: String, optionId: String) async throws {
        struct VotePollBody: Encodable {
            let optionId: String
        }
        let _: EmptyResponse = try await http.post("/meetings/\(meetingId)/polls/\(pollId)/vote", body: VotePollBody(optionId: optionId))
    }

    public func createBreakoutRooms(_ meetingId: String, rooms: [BreakoutRoomInput]) async throws -> [BreakoutRoom] {
        struct CreateBreakoutRoomsBody: Encodable {
            let rooms: [BreakoutRoomInput]
        }
        let result: [BreakoutRoom] = try await http.post("/meetings/\(meetingId)/breakout-rooms", body: CreateBreakoutRoomsBody(rooms: rooms))
        return result
    }

    public func raiseHand(_ meetingId: String) {
        Task { await ws.send(type: "hand.raise", data: ["meetingId": meetingId]) }
    }

    public func lowerHand(_ meetingId: String) {
        Task { await ws.send(type: "hand.lower", data: ["meetingId": meetingId]) }
    }

    /// Async stream of participant events (joined/left) for a given meeting.
    public func observeParticipants(meetingId: String) -> AsyncStream<MeetingParticipant> {
        AsyncStream { continuation in
            Task {
                await ws.on("meeting.participant.joined") { data in
                    if let p = try? JSONDecoder().decode(MeetingParticipant.self, from: data),
                       p.meetingId == meetingId {
                        continuation.yield(p)
                    }
                }
            }
        }
    }

    /// Async stream of poll events for a given meeting.
    public func observePolls(meetingId: String) -> AsyncStream<MeetingPoll> {
        AsyncStream { continuation in
            Task {
                await ws.on("meeting.poll.created") { data in
                    if let poll = try? JSONDecoder().decode(MeetingPoll.self, from: data),
                       poll.meetingId == meetingId {
                        continuation.yield(poll)
                    }
                }
                await ws.on("meeting.poll.result") { data in
                    if let poll = try? JSONDecoder().decode(MeetingPoll.self, from: data),
                       poll.meetingId == meetingId {
                        continuation.yield(poll)
                    }
                }
            }
        }
    }
}

public struct MeetingPoll: Codable {
    public let id: String
    public let meetingId: String
    public let question: String
    public let options: [PollOption]
    public let isActive: Bool
    public let isAnonymous: Bool
}

public struct PollOption: Codable {
    public let id: String
    public let text: String
    public let votes: Int
}

public struct BreakoutRoom: Codable {
    public let id: String
    public let name: String
    public let meetingId: String
    public let participants: [String]
}

public struct BreakoutRoomInput: Encodable {
    public let name: String
    public let participants: [String]
    
    public init(name: String, participants: [String]) {
        self.name = name
        self.participants = participants
    }
}

public struct MeetingConfigInput: Encodable {
    public let title: String
    public let description: String?
    public let scheduledAt: String?
    public let durationMins: Int

    public init(title: String, description: String? = nil, scheduledAt: String? = nil, durationMins: Int = 60) {
        self.title = title
        self.description = description
        self.scheduledAt = scheduledAt
        self.durationMins = durationMins
    }
}

File: sdks/android/streamkit-sdk/src/main/kotlin/com/rajutechie/streamkit/modules/MeetingModule.kt
package com.rajutechie.streamkit.modules

import com.rajutechie.streamkit.transport.HttpClient
import com.rajutechie.streamkit.transport.WebSocketClient
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.Flow

data class Meeting(val id: String, val title: String, val meetingCode: String, val status: String, val hostId: String)
data class MeetingParticipant(val id: String, val meetingId: String, val userId: String, val role: String, val status: String)
data class MeetingConfig(val title: String, val description: String? = null, val scheduledAt: String? = null, val durationMins: Int = 60)
data class MeetingPoll(val id: String, val question: String, val options: List<PollOption>, val isActive: Boolean)
data class PollOption(val id: String, val text: String)

class MeetingModule(
    private val http: HttpClient,
    private val ws: WebSocketClient,
    private val scope: CoroutineScope
) {
    suspend fun schedule(config: MeetingConfig): Meeting = http.post("/meetings", config)
    suspend fun get(meetingId: String): Meeting = http.get("/meetings/$meetingId")
    suspend fun join(meetingId: String): MeetingParticipant = http.post("/meetings/$meetingId/join")
    suspend fun joinByCode(code: String): Meeting = http.get("/meetings/join/$code")
    suspend fun leave(meetingId: String) { http.post<Unit>("/meetings/$meetingId/leave") }
    suspend fun end(meetingId: String) { http.post<Unit>("/meetings/$meetingId/end") }
    suspend fun muteParticipant(meetingId: String, userId: String) { http.post<Unit>("/meetings/$meetingId/participants/$userId/mute") }
    suspend fun removeParticipant(meetingId: String, userId: String) { http.post<Unit>("/meetings/$meetingId/participants/$userId/remove") }
    suspend fun muteAll(meetingId: String) { http.post<Unit>("/meetings/$meetingId/mute-all") }

    fun raiseHand(meetingId: String) { ws.emit("hand.raise", mapOf("meetingId" to meetingId)) }
    fun lowerHand(meetingId: String) { ws.emit("hand.lower", mapOf("meetingId" to meetingId)) }

    suspend fun createPoll(meetingId: String, question: String, options: List<String>): MeetingPoll {
        return http.post("/meetings/$meetingId/polls", mapOf("question" to question, "options" to options))
    }

    suspend fun votePoll(meetingId: String, pollId: String, optionId: String) {
        http.post<Unit>("/meetings/$meetingId/polls/$pollId/vote", mapOf("optionId" to optionId))
    }

    suspend fun createBreakoutRooms(meetingId: String, rooms: List<Map<String, Any>>): List<BreakoutRoom> {
        return http.post("/meetings/$meetingId/breakout-rooms", mapOf("rooms" to rooms))
    }

    fun observeParticipants(): Flow<MeetingParticipant> = ws.on("meeting.participant.joined", MeetingParticipant::class.java)

    fun observePolls(): Flow<MeetingPoll> = ws.on("meeting.poll.created", MeetingPoll::class.java)

    fun observeParticipantLeft(): Flow<Map<String, String>> = ws.on("meeting.participant.left", Map::class.java as Class<Map<String, String>>)
}

data class BreakoutRoom(val id: String, val name: String, val meetingId: String, val participants: List<String>)
